src/genh5_dataset.cpp

```diff
@@ -16,25 +16,86 @@
 #include "H5Dpublic.h"
 #include "H5Ppublic.h"
 
-GenH5::DataSet::DataSet() = default;
+#include <QDebug>
+
+struct GenH5::DataSet::Impl
+{
+    explicit Impl(hid_t id)
+        : id(id)
+    {}
+
+    /// dataset id
+    IdComponent<H5I_DATASET> id;
+
+    // The hooks to execute before and after read / write
+    std::map<GenH5::HookType, GenH5::Hook> hooks;
+};
+
+GenH5::DataSet::DataSet()
+    : GenH5::DataSet(hid_t())
+{}
 
 GenH5::DataSet::DataSet(hid_t id) :
-    m_id(id)
+    pimpl(std::make_unique<Impl>(id))
+{
+    pimpl->id.inc();
+}
+
+GenH5::DataSet::DataSet(const DataSet & other)
+    : pimpl(std::make_unique<Impl>(*other.pimpl))
+{
+}
+
+GenH5::DataSet::DataSet(DataSet && other)
+    : pimpl(std::move(other.pimpl))
+{
+}
+
+GenH5::DataSet&
+GenH5::DataSet::operator=(const DataSet& other)
 {
-    m_id.inc();
+    GenH5::DataSet tmp(other);
+    swap(tmp);
+    return *this;
 }
 
+GenH5::DataSet&
+GenH5::DataSet::operator=(DataSet&& other)
+{
+    swap(other);
+    return *this;
+}
+
+
+GenH5::DataSet::~DataSet() = default;
+
 void
 GenH5::DataSet::swap(DataSet& other) noexcept
 {
     using std::swap;
-    swap(m_id, other.m_id);
+    swap(pimpl, other.pimpl);
+}
+
+void
+GenH5::DataSet::setHook(const GenH5::Hook &hook, GenH5::HookType type)
+{
+    pimpl->hooks[type] = hook;
+}
+
+void GenH5::DataSet::execHook(GenH5::HookType type) const
+{
+    auto iter = pimpl->hooks.find(type);
+    if (iter != pimpl->hooks.end() && iter->second)
+    {
+        // execute
+        iter->second();
```

Comment:
Maybe you could extend it further by passing the dataset its called on as a parameter. Then you may not need a unique hook per dataset? But I don't know how you use this API exactly.

```suggestion
        iter->second(*this);
```

Reply:
I have also thought about this. Lets put our heads together and think about a good solution. the hooks must be non-static though, they need to depend on a specific file-id, since some files might be local, other might be remote.

Reply:
Maybe, we could register hooks per file, not per dataset. I.e. dataset hooks etc. It is probably not required to have a different hook for each dataset within the same file

Reply:
But should the hooks be unique for the file (i.e. "physical" file/filepath) or should they be unique per hdf5-file-instance? I.e. we could open the same file multiple time, each handle would be allowed to have different hooks. What do you think? @rainman110
I think the latter would be more intuitive

Reply:
File instance obviously. What i mean is that the hooks could be placed inside the h5file object.

Reply:
Sounds good!

Reply:
I implemented file-wide hooks in the newest commit